dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java

```diff
@@ -82,8 +124,63 @@ protected EPersonRest createAndReturn(Context context)
+        RegistrationData registrationData = registrationDataService.findByToken(context, token);
+        if (registrationData == null) {
+            throw new DSpaceBadRequestException("The token given as parameter: " + token + " does not exist" +
+                                                " in the database");
+        }
+        if (es.findByEmail(context, registrationData.getEmail()) != null) {
+            throw new DSpaceBadRequestException("The token given already contains an email address that resolves" +
+                                                "to an eperson");
+        }
+        String emailFromJson = epersonRest.getEmail();
+        if (StringUtils.isNotBlank(emailFromJson)) {
+            if (!StringUtils.equalsIgnoreCase(registrationData.getEmail(), emailFromJson)) {
+                throw new DSpaceBadRequestException("The email resulting from the token does not match the email given"
+                                                        + " in the json body. Email from token: " +
+                                                    registrationData.getEmail() + " email from the json body: "
+                                                    + emailFromJson);
+            }
+        }
+        if (epersonRest.isSelfRegistered() != null && !epersonRest.isSelfRegistered()) {
+            throw new DSpaceBadRequestException("The self registered property cannot be set to false using this method"
+                                                    + " with a token");
+        }
+        checkRequiredProperties(epersonRest);
+        if (!AuthorizeUtil.authorizeNewAccountRegistration(context, requestService
+            .getCurrentRequest().getHttpServletRequest())) {
+            throw new DSpaceBadRequestException(
+                "Registration is disabled, you are not authorized to create a new Authorization");
+        }
+        // We'll turn off authorisation system because this call isn't admin based as it's token based
+        context.turnOffAuthorisationSystem();
+        EPerson ePerson = createEPersonFromRestObject(context, epersonRest);
+        context.restoreAuthSystemState();
+        // Restoring authorisation state right after the creation call
+        accountService.deleteToken(context, token);
+        return converter.toRest(ePerson, utils.obtainProjection());
+    }
 
-        return converter.toRest(eperson, utils.obtainProjection());
+    private void checkRequiredProperties(EPersonRest epersonRest) {
+        MetadataRest metadataRest = epersonRest.getMetadata();
+        if (metadataRest != null) {
+            List<MetadataValueRest> epersonFirstName = metadataRest.getMap().get("eperson.firstname");
+            List<MetadataValueRest> epersonLastName = metadataRest.getMap().get("eperson.lastname");
+            if (epersonFirstName == null || epersonLastName == null ||
+                epersonFirstName.isEmpty() || epersonLastName.isEmpty()) {
```

Comment:
This if can be simplified by using `StringUtils.isNotBlank` as you have in other methods. It checks for both null and empty

Reply:
epersonFirstName and epersonLastName is not a String, but rather `List<MetadataValueRest>`. The StringUtils.isNotBlank doesn't support that AFAIK